src/main/java/org/jabref/logic/importer/PagedSearchBasedFetcher.java

```diff
@@ -0,0 +1,26 @@
+package org.jabref.logic.importer;
+
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.paging.Page;
+
+public interface PagedSearchBasedFetcher extends SearchBasedFetcher {
```

Comment:
Could you add some JavaDoc here about the idea?

I see that one can

a) jump to arbitrary pages
b) the page size supported is pre-defined

Would there also be a method `setPageSize()` make sense? Maybe a sub interface `PagedSearchBasedFetcherWithVariablePageSize`?

The alternative solution could have been `iterate` as [jcabi github](https://github.jcabi.com/apidocs-0.41/index.html) does it. However, with your approach, one can directly jump to a certain page and is not forced to query sequentially.

Reply:
I tried to keep the fetcher as stateless as possible. If thats is not required, i can provide an interface for that purpose.
It probably depends where we want configure the page size. In the application settings or as a dropdown?

Reply:
I would like to see it in a drop down. Internally, it can be persisted as preference, so that at a restart of JabRef, the value is the same.

I can imagine that with pagination we need to be stateful. If I as user go back and forth in the UI, I want to have quick responses. If there is a request sent to the server each time, this will be slow.

I am aware that this will require a "refresh search results" button in the UI, which just resets the factory to trigger new searches.

I would even use [Google Guava Cache](https://seeseekey.net/archive/122813/) for caching different search results over different fetchers. -- The search results for publications won't differ from hour to hour, but merely from day to day or even week to week. Having the possibility to limit the cache by memory size. - I also find the loading pretty nice. see https://github.com/google/guava/wiki/CachesExplained#from-a-callable:

```java
CacheLoader<Key, Graph> loader = CacheLoader.from(key -> createExpensiveGraph(key));
```

Source: https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/CacheLoader.html

WDYT?

Reply:
I would let each fetcher decide the appropriate page size. Some fetchers might not support a variable page size, and even if they do there might be restrictions (for example some fetchers get a list of ids and then fetch details for every id; in this case you probably want to have only a small page size).

Reply:
> I would like to see it in a drop down. Internally, it can be persisted as preference, so that at a restart of JabRef, the value is the same.

Then I think we dont need a setSize method, because how the getSize is implemented is up to the fetcher. The fetcher can then use the Preference API.

 > I can imagine that with pagination we need to be stateful. If I as user go back and forth in the UI, I want to have quick responses. If there is a request sent to the server each time, this will be slow.
I would argue that the fetcher doesn't need to be stateless but should be stateless. It's solely responsibility is to fetch results. The ViewModel can then cache the results. Caching is possible in the fetcher as well, but the Viewmodel is statefull anyway, therefore we don't add too much new complexity.

Reply:
>I would even use Google Guava Cache for caching different search results over different fetchers. -- The search results for publications won't differ from hour to hour, but merely from day to day or even week to week. Having the possibility to limit the cache by memory size. - I also find the loading pretty nice. see https://github.com/google/guava/wiki/CachesExplained#from-a-callable:

That's another reason to do it in the viewmodel

Reply:
Maybe I can implement a cache decorator for the fetchers. That way caching becomes very easy.

Reply:
I was thinking, that query optimization can be done best at the result provider. Thus, if I query 1000 elements at once as JabRef, following can happen:

1. Provider supports 1000 entrie at once --> they are returned -(query optimization there)
2. Provider supports 50 entries at once --> jabref does 20 calls to the provider (nearly no query optimization at the provider)

My thoughts with the factory were driven with the idea that the first setting should be supported.

That leads me to the idea that we might need `maxPageSize` and `pageSize`. Maybe even different page sizes for the view and for the fetcher...

We can surely separate the the basic fetcher and some layer inbetween.

UI --> caching layer --> fetcher

Reply:
Caching is a nice idea, but not thaaaat important. I would say, first focus on the paging UI and then we can have a look at how to implement caching.